### PR TITLE
Support for non-empty arrays as defaults in migrations

### PR DESCRIPTION
**Pull request by sechosebio**
Support for non-empty arrays as defaults in migrations as requested on #2482. 
Default arrays can be string arrays and integer arrays.

### CODE REVIEW

File: lib/ecto/adapters/postgres/connection.ex [hunk 1]
```diff
@@ -808,6 +808,15 @@ if Code.ensure_loaded?(Postgrex) do
       do: " DEFAULT NULL"
     defp default_expr({:ok, []}, type),
       do: [" DEFAULT ARRAY[]::", ecto_to_db(type)]
+    defp default_expr({:ok, list}, {:array, :string} = type) when is_list(list) do
+      [" DEFAULT ARRAY['#{Enum.join(list, "', '")}']::", ecto_to_db(type)]
```
Comment (ericmj) [inline lib/ecto/adapters/postgres/connection.ex]: The strings inside the list need to be escaped.

**Comment by josevalim** [comment]
Thanks @sechosebio! It would probably be best if we were able to call default recursively, instead of repeating the clauses In pseudo code it would look like this:

```elixir
defp default_expr({:ok, value}, type) do
  [" DEFAULT ", default_type(value)]
end

defp default_expr(:error, _) do
  []
end

defp default_type(list, {:array, inner} = type) when is_list(list) do
  ["ARRAY[", Enum.map(list, &default_type(&1, inner)) |> Enum.intersperse(?,), "]", ecto_to_db(type)]
end

defp default_type(string, ...)
defp default_type(number, ...)
defp default_type(boolean, ...)
```

This way we don't need to repeat string, number and so on handling everywhere.

**Comment by josevalim** [comment]
This is beautiful! Can you please apply those changes to the MySQL adapter too and add a test? Then this should be ready to go! Running `mix test` will be enough to test the generation of SQL for both Postgres and MySQL.

**Comment by sechosebio** [comment]
@josevalim MySQL doesn't support the array type yet, at least not in the way that Postgres does. Or do you mean refactor the code as in the Postgres adapter? 🤔

**Comment by josevalim** [comment]
@sechosebio you are right. I guess there is no point in refactoring it if it doesn't support array types.

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Pull request merged**